File: sloth/views/reports.py
"""Saved nmap reports, screenshots, and exports.

The old export dumped whatever masscan happened to have in memory. These pull
from the database instead, so an export covers everything a task or project ever
found: masscan ports, nmap-detected services, the raw nmap output and the
screenshots, embedded so the HTML file stands alone.
"""
import base64
import json
import os

from flask import (Blueprint, abort, flash, jsonify, make_response, redirect,
                   render_template, request, send_file, url_for)

from .. import store, transfer
from ..config import SHOTS_DIR

bp = Blueprint("reports", __name__)


@bp.route("/screenshot/<path:fname>")
def screenshot(fname):
    # basename() keeps a crafted name from escaping the screenshots directory.
    path = os.path.join(SHOTS_DIR, os.path.basename(fname))
    if not os.path.isfile(path):
        abort(404)
    return send_file(path, mimetype="image/png")


SCANS_PAGE_SIZE = 200


@bp.route("/scans")
def list_scans():
    # Search runs in SQL so it reaches every saved scan, not just the page shown.
    search = (request.args.get("q") or "").strip()
    project_id = (request.args.get("project") or "").strip() or None
    matched = store.list_nmap_scans(project_id=project_id, search=search or None)

    limit = SCANS_PAGE_SIZE
    if request.args.get("all"):
        limit = len(matched)

    # Only projects that actually have a report — offering empty ones is just a
    # list of dead ends.
    projects = [p for p in store.list_projects()
                if store.list_nmap_scans(project_id=p["id"])]

    return render_template(
        "scans.html", nav_section="scans", scans=matched[:limit],
        total=len(matched), truncated=len(matched) > limit,
        grand_total=len(store.list_nmap_scans()),
        search=search, project_id=project_id or "", projects=projects)


@bp.route("/nmap-result/<scan_id>")
def nmap_result(scan_id):
    row = store.get_nmap_scan(scan_id)
    if row is None:
        abort(404)
    return render_template(
        "nmap_result.html",
        nav_section="scans",
        scan=row,
        ports=_loads(row["ports_json"]),
        shots=_loads(row["screenshots_json"]),
    )


@bp.route("/export/task/<task_id>")
def export_task(task_id):
    task = store.get_task(task_id)
    if task is None:
        abort(404)
    bundle = _task_bundle(task)
    return _render_export(request.args.get("format", "html"),
                          title=f"{task['project_name']} · {task['name']}",
                          slug=f"task-{task_id}",
                          sections=[bundle])


@bp.route("/export/project/<project_id>")
def export_project(project_id):
    project = store.get_project(project_id)
    if project is None:
        abort(404)
    sections = [_task_bundle(t) for t in store.list_tasks(project_id)]
    return _render_export(request.args.get("format", "html"),
                          title=project["name"],
                          slug=f"project-{project_id}",
                          sections=sections,
                          project=project)


@bp.route("/import", methods=["POST"])
def import_bundle():
    """Loads a JSON export from another install.

    Always lands in a project — a new one by default, or an existing one when
    two people are splitting the same engagement between them.
    """
    back = request.form.get("project_id")
    home = (url_for("projects.project_detail", project_id=back) if back
            else url_for("projects.index"))

    upload = request.files.get("bundle")
    if upload is None or not upload.filename:
        flash("Choose a .json export to import.", "error")
        return redirect(home)

    try:
        data = transfer.read_bundle(upload.read())
        tally = transfer.import_bundle(data, project_id=back or None)
    except transfer.BundleError as exc:
        flash(str(exc), "error")
        return redirect(home)

    counts = " · ".join(
        f"{tally[k]} {label}"
        for k, label in (("tasks", "task(s)"), ("hosts", "host(s)"),
                         ("ports", "port(s)"), ("scans", "nmap scan(s)"),
                         ("shots", "screenshot(s)")) if tally[k])
    where = "into a new project" if tally["created"] else "into this project"
    flash(f"Imported {counts or 'nothing — the file had no results'} {where}.",
          "ok")
    if tally["skipped"]:
        flash(f"{tally['skipped']} entr(y/ies) in the file were malformed and "
              f"were skipped.", "error")
    return redirect(url_for("projects.project_detail",
                            project_id=tally["project"]))


def _task_bundle(task):
    """Everything one task knows, ready for rendering or serialising."""
    hosts = store.task_hosts(task["id"])
    scans = {}
    for row in store.list_nmap_scans(task_id=task["id"]):
        full = store.get_nmap_scan(row["id"])
        scans.setdefault(row["ip"], []).append({
            "id": full["id"],
            "tool": full["tool"],
            "created_at": full["created_at"],
            "command": full["command"],
            "raw_output": full["raw_output"],
            "screenshots": _loads(full["screenshots_json"]),
        })
    return {"task": dict(task), "hosts": hosts, "scans": scans}


def _render_export(fmt, title, slug, sections, project=None):
    if fmt == "json":
        # The JSON export is also the import format, so it carries a version
        # header and the screenshots themselves rather than filenames that mean
        # nothing on another machine.
        from .. import __version__
        payload = transfer.envelope(title, sections, project, version=__version__)
        resp = make_response(json.dumps(payload, indent=2, default=str))
        resp.headers["Content-Type"] = "application/json"
        resp.headers["Content-Disposition"] = f"attachment; filename={slug}.json"
        return resp

    if fmt == "txt":
        resp = make_response(_render_text(title, sections, project))
        resp.headers["Content-Type"] = "text/plain; charset=utf-8"
        resp.headers["Content-Disposition"] = f"attachment; filename={slug}.txt"
        return resp

    # HTML: self-contained, screenshots inlined as data URIs.
    for sec in sections:
        for scan_list in sec["scans"].values():
            for scan in scan_list:
                for shot in scan["screenshots"]:
                    shot["data_uri"] = _inline_png(shot.get("file"))
    html = render_template("export.html", title=title, sections=sections,
                           project=project)
    resp = make_response(html)
    resp.headers["Content-Type"] = "text/html; charset=utf-8"
    resp.headers["Content-Disposition"] = f"attachment; filename={slug}.html"
    return resp


def _scan_config(t):
    """The one-line description of how a task was run, as the pages show it."""
    bits = [t["scan_type"] or "full"]
    for label, key in (("", "engine"), ("discovery ", "discovery"),
                       ("TCP ", "tcp_ports"), ("UDP ", "udp_ports"),
                       ("top ", "top_ports")):
        if t[key]:
            bits.append(f"{label}{t[key]}")
    if t["rate"]:
        bits.append(f"{t['rate']} pkts/s")
    return " · ".join(bits)


def _render_text(title, sections, project):
    """Plain text, carrying what the HTML report carries.

    The finding lines keep the shape 'ip:port (proto/state)  service' — this is
    the format people pipe into grep and cut, so it stays stable even as the
    surrounding report grows.
    """
    out = [f"# {title}"]
    if project:
        if project["client"]:
            out.append(f"client: {project['client']}")
        out.append(f"project created {project['created_at']}")
        if project["description"]:
            out.append(project["description"])

    hosts = sum(len(s["hosts"]) for s in sections)
    ports = sum(len(h["ports"]) for s in sections for h in s["hosts"])
    out += ["", f"{len(sections)} task(s) · {hosts} host(s) · "
                f"{ports} open/filtered port(s)"]

    for sec in sections:
        t = sec["task"]
        out += ["", "=" * 72, f"## {t['name']}", ""]
        meta = [("target", t["target"]), ("status", t["status"]),
                ("scan", _scan_config(t)), ("started", t["started_at"]),
                ("finished", t["finished_at"]), ("notes", t["notes"]),
                ("error", t["error"])]
        out += [f"   {k:<10} {v}" for k, v in meta if v]
        out.append("")

        if not sec["hosts"]:
            out.append("   (no hosts with open ports were found by this task)")
            continue

        for host in sec["hosts"]:
            if not host["ports"]:
                out.append(f"{host['ip']}  (up — no open ports recorded)")
            for p in host["ports"]:
                svc = f"  {p['service']}" if p.get("service") else ""
                out.append(
                    f"{host['ip']}:{p['port']} ({p['proto']}/{p['state']}){svc}")

            for scan in sec["scans"].get(host["ip"], []):
                out += ["", f"   nmap · {scan['tool']} · {scan['created_at']}",
                        f"   $ {scan['command']}"]
                for shot in scan["screenshots"]:
                    out.append(f"   screenshot: {shot.get('url', '')}")
                out += ["   " + "-" * 60]
                out += ["   " + ln for ln in
                        (scan["raw_output"] or "").rstrip().splitlines()]
                out += ["   " + "-" * 60, ""]
    return "\n".join(out) + "\n"


def _inline_png(fname):
    if not fname:
        return None
    path = os.path.join(SHOTS_DIR, os.path.basename(fname))
    try:
        with open(path, "rb") as fh:
            return "data:image/png;base64," + base64.b64encode(fh.read()).decode()
    except OSError:
        return None


def _loads(value):
    try:
        return json.loads(value or "[]")
    except (ValueError, TypeError):
        return []

File: src/main/reports.js
'use strict'
/** Exports: the self-contained HTML report, plain text, and the JSON bundle.
 *
 * Ported from views/reports.py and templates/export.html. The Jinja template
 * becomes a function here; the markup, the inlined Nocturne palette and the
 * decision not to embed 4 MB of webfonts are all unchanged, so a report from
 * this build looks the same as one from the Python build.
 */
const fs = require('node:fs')
const path = require('node:path')

const store = require('./store')
const transfer = require('./transfer')
const { SHOTS_DIR } = require('./config')

/** Everything one task knows, ready for rendering or serialising. */
function taskBundle (task) {
  const hosts = store.taskHosts(task.id)
  const scans = {}
  for (const row of store.listNmapScans({ taskId: task.id })) {
    const full = store.getNmapScan(row.id)
    if (!scans[full.ip]) scans[full.ip] = []
    scans[full.ip].push({
      id: full.id,
      tool: full.tool,
      created_at: full.created_at,
      command: full.command,
      raw_output: full.raw_output,
      screenshots: parseJson(full.screenshots_json),
    })
  }
  return { task: { ...task }, hosts, scans }
}

function parseJson (value) {
  try {
    const parsed = JSON.parse(value || '[]')
    return Array.isArray(parsed) ? parsed : []
  } catch {
    return []
  }
}

/** Renders one export. Returns {filename, mime, body}. */
function render (format, { title, slug, sections, project = null, version = '?' }) {
  if (format === 'json') {
    const payload = transfer.envelope(title, sections, project, version)
    return {
      filename: `${slug}.json`,
      mime: 'application/json',
      body: JSON.stringify(payload, null, 2),
    }
  }
  if (format === 'txt') {
    return {
      filename: `${slug}.txt`,
      mime: 'text/plain; charset=utf-8',
      body: renderText(title, sections, project),
    }
  }
  // HTML: self-contained, screenshots inlined as data URIs.
  for (const section of sections) {
    for (const list of Object.values(section.scans)) {
      for (const scan of list) {
        for (const shot of scan.screenshots) shot.data_uri = inlinePng(shot.file)
      }
    }
  }
  return {
    filename: `${slug}.html`,
    mime: 'text/html; charset=utf-8',
    body: renderHtml(title, sections, project),
  }
}

function inlinePng (name) {
  if (!name) return null
  try {
    const raw = fs.readFileSync(path.join(SHOTS_DIR, path.basename(name)))
    return 'data:image/png;base64,' + raw.toString('base64')
  } catch {
    return null
  }
}

/** The one-line description of how a task was run, as the pages show it. */
function scanConfig (task) {
  const bits = [task.scan_type || 'full']
  for (const [label, key] of [['', 'engine'], ['discovery ', 'discovery'],
    ['TCP ', 'tcp_ports'], ['UDP ', 'udp_ports'], ['top ', 'top_ports']]) {
    if (task[key]) bits.push(`${label}${task[key]}`)
  }
  if (task.rate) bits.push(`${task.rate} pkts/s`)
  return bits.join(' · ')
}

/** Plain text, carrying what the HTML report carries.
 *
 * The finding lines keep the shape 'ip:port (proto/state)  service' — this is
 * the format people pipe into grep and cut, so it stays stable even as the
 * surrounding report grows.
 */
function renderText (title, sections, project) {
  const out = [`# ${title}`]
  if (project) {
    if (project.client) out.push(`client: ${project.client}`)
    out.push(`project created ${project.created_at}`)
    if (project.description) out.push(project.description)
  }

  const hosts = sections.reduce((n, s) => n + s.hosts.length, 0)
  const ports = sections.reduce((n, s) =>
    n + s.hosts.reduce((m, h) => m + h.ports.length, 0), 0)
  out.push('', `${sections.length} task(s) · ${hosts} host(s) · ${ports} open/filtered port(s)`)

  for (const section of sections) {
    const task = section.task
    out.push('', '='.repeat(72), `## ${task.name}`, '')
    const meta = [['target', task.target], ['status', task.status],
      ['scan', scanConfig(task)], ['started', task.started_at],
      ['finished', task.finished_at], ['notes', task.notes], ['error', task.error]]
    for (const [key, value] of meta) {
      if (value) out.push(`   ${key.padEnd(10)} ${value}`)
    }
    out.push('')

    if (!section.hosts.length) {
      out.push('   (no hosts with open ports were found by this task)')
      continue
    }

    for (const host of section.hosts) {
      if (!host.ports.length) out.push(`${host.ip}  (up — no open ports recorded)`)
      for (const port of host.ports) {
        const service = port.service ? `  ${port.service}` : ''
        out.push(`${host.ip}:${port.port} (${port.proto}/${port.state})${service}`)
      }
      for (const scan of section.scans[host.ip] || []) {
        out.push('', `   nmap · ${scan.tool} · ${scan.created_at}`, `   $ ${scan.command}`)
        for (const shot of scan.screenshots) out.push(`   screenshot: ${shot.url || ''}`)
        out.push('   ' + '-'.repeat(60))
        for (const line of String(scan.raw_output || '').replace(/\s+$/, '').split('\n')) {
          out.push('   ' + line)
        }
        out.push('   ' + '-'.repeat(60), '')
      }
    }
  }
  return out.join('\n') + '\n'
}

/** Escapes text for HTML. Everything interpolated below is attacker-influenced
 *  — service banners, hostnames and nmap output come off the scanned host. */
function esc (value) {
  if (value === null || value === undefined) return ''
  return String(value)
    .replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;').replace(/'/g, '&#39;')
}

const STYLE = `
    /* Nocturne's palette, inlined. Deliberately self-contained: no stylesheet
       link, no CDN, screenshots embedded as data URIs, so this file still
       renders years from now on a machine with no network.

       The webfonts are the one thing left out — Inter and the icon set are
       4 MB, and attaching that to every exported report to change the shape of
       the letters is a poor trade. System stacks stand in. */
    :root {
      --bg:        oklch(0.212 0.033 255);
      --surface:   oklch(0.276 0.026 255);
      --sunken:    oklch(0.259 0.018 255);
      --neutral:   oklch(0.355 0.024 255);
      --text:      #e9e9ed;
      --accent:    oklch(0.660 0.125 258);
      --accent-300:oklch(0.859 0.055 258);
      --divider:   color-mix(in srgb, #e9e9ed 16%, transparent);
      --muted:     color-mix(in srgb, #e9e9ed 55%, transparent);
      --faint:     color-mix(in srgb, #e9e9ed 40%, transparent);
      --mono:      ui-monospace, SFMono-Regular, Menlo, Consolas, monospace;
    }
    * { box-sizing: border-box; }
    body { background: var(--bg); color: var(--text); margin: 0; padding: 34px 22px 80px;
           font-family: system-ui, -apple-system, "Segoe UI", Roboto, sans-serif;
           font-size: 14px; line-height: 1.55; }
    .wrap { max-width: 1080px; margin: 0 auto; }

    h1 { font-size: 24px; letter-spacing: -0.02em; margin: 0 0 6px; }
    h2 { font-size: 17px; margin: 40px 0 4px; }
    h3 { font-size: 12px; letter-spacing: 0.08em; text-transform: uppercase;
         color: var(--accent-300); margin: 22px 0 8px; font-weight: 600; }
    .kicker { font-size: 11px; letter-spacing: 0.08em; text-transform: uppercase;
              color: var(--faint); margin: 0 0 6px; }
    .meta  { color: var(--muted); font-size: 12px; margin: 4px 0 0; }
    .head  { padding-bottom: 17px; margin-bottom: 22px; border-bottom: 1px solid var(--divider); }
    .empty { color: var(--faint); font-style: italic; font-size: 12px; }

    .card { background: var(--surface); border: 1px solid var(--divider);
            border-radius: 14px; padding: 17px; margin: 11px 0 17px; }
    .ip   { font-family: var(--mono); font-size: 15px; font-weight: 600; }

    table { width: 100%; border-collapse: collapse; font-size: 13px; margin-top: 11px; }
    th { text-align: left; font-size: 11px; letter-spacing: 0.08em; text-transform: uppercase;
         color: var(--muted); font-weight: 600; padding: 6px 8px;
         border-bottom: 1px solid var(--divider); }
    td { padding: 6px 8px; border-bottom: 1px solid
         color-mix(in srgb, #e9e9ed 8%, transparent); vertical-align: top; }
    td.port { font-family: var(--mono); color: var(--accent-300); white-space: nowrap; }
    td.src  { font-size: 11px; color: var(--faint); }

    /* One neutral pill, matching the interface: state is text, not alarm colour. */
    .tag { display: inline-block; font-size: 11px; padding: 2px 8px; border-radius: 999px;
           background: var(--neutral); border: 1px solid var(--divider);
           font-family: var(--mono); white-space: nowrap; }
    .tag.dim { opacity: 0.62; }

    pre { white-space: pre-wrap; word-break: break-word; background: var(--sunken);
          border: 1px solid var(--divider); border-radius: 8px; padding: 12px;
          font-family: var(--mono); font-size: 11.5px; line-height: 1.65;
          color: color-mix(in srgb, #e9e9ed 72%, transparent); overflow-x: auto; margin: 0; }
    details { margin: 10px 0 0; }
    summary { cursor: pointer; color: var(--accent-300); font-size: 12px; }

    .shots { display: grid; grid-template-columns: repeat(auto-fit, minmax(320px, 1fr));
             gap: 14px; margin-top: 8px; }
    img { max-width: 100%; display: block; border: 1px solid var(--divider);
          border-radius: 8px; }

    /* Reports get printed and pasted into deliverables. */
    @media print {
      body { background: #fff; color: #111; }
      .card, pre { background: #fff; border-color: #ccc; }
      .tag { background: #eee; }
      a { color: inherit; }
    }
`

function renderHtml (title, sections, project) {
  const hostTotal = sections.reduce((n, s) => n + s.hosts.length, 0)
  const portTotal = sections.reduce((n, s) =>
    n + s.hosts.reduce((m, h) => m + h.ports.length, 0), 0)

  const parts = []
  parts.push(`<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>${esc(title)} — Sloth report</title>
<style>${STYLE}</style>
</head>
<body>
<div class="wrap">
    <div class="head">
        <p class="kicker">Sloth report</p>
        <h1>${esc(title)}</h1>
        <p class="meta">`)
  if (project) {
    if (project.client) parts.push(`client: ${esc(project.client)} · `)
    parts.push(`project created ${esc(project.created_at)}`)
  }
  parts.push('</p>')
  if (project && project.description) {
    parts.push(`\n        <p class="meta">${esc(project.description)}</p>`)
  }
  parts.push(`
        <p class="meta">
            ${sections.length} task(s) · ${hostTotal} host(s) ·
            ${portTotal} open/filtered port(s)
        </p>
    </div>
`)

  for (const section of sections) {
    const task = section.task
    const meta = [
      `${esc(task.scan_type || 'full')}`,
      task.engine ? ` · ${esc(task.engine)}` : '',
      ` · target ${esc(task.target)} · status ${esc(task.status)}`,
      task.discovery ? ` · discovery ${esc(task.discovery)}` : '',
      task.tcp_ports ? ` · TCP ${esc(task.tcp_ports)}` : '',
      task.udp_ports ? ` · UDP ${esc(task.udp_ports)}` : '',
      task.top_ports ? ` · top ${esc(task.top_ports)}` : '',
      task.rate ? ` · ${esc(task.rate)} pkts/s` : '',
      task.started_at ? ` · started ${esc(task.started_at)}` : '',
      task.finished_at ? ` · finished ${esc(task.finished_at)}` : '',
    ].join('')

    parts.push(`    <h2>${esc(task.name)}</h2>\n    <p class="meta">${meta}</p>\n`)
    if (task.notes) parts.push(`    <p class="meta">Notes: ${esc(task.notes)}</p>\n`)
    if (task.error) parts.push(`    <p class="meta">Error: ${esc(task.error)}</p>\n`)

    if (!section.hosts.length) {
      parts.push('    <p class="empty" style="margin-top:11px">No hosts with open ' +
        'ports were found by this task.</p>\n')
    }

    for (const host of section.hosts) {
      parts.push(`    <div class="card">\n        <div class="ip">${esc(host.ip)}</div>\n`)
      if (host.ports.length) {
        parts.push(`        <table>
            <thead><tr>
                <th style="width:110px">Port</th><th style="width:130px">State</th>
                <th>Service</th><th style="width:90px">Source</th>
            </tr></thead>
            <tbody>
`)
        for (const port of host.ports) {
          const dim = String(port.state || '').startsWith('open') ? '' : ' dim'
          parts.push(`            <tr>
                <td class="port">${esc(port.port)}/${esc(port.proto)}</td>
                <td><span class="tag${dim}">${esc(port.state)}</span></td>
                <td>${esc(port.service || '—')}</td>
                <td class="src">${esc(port.source)}</td>
            </tr>
`)
        }
        parts.push('            </tbody>\n        </table>\n')
      } else {
        parts.push('        <p class="empty" style="margin-top:8px">host is up — ' +
          'no open ports recorded</p>\n')
      }

      for (const scan of section.scans[host.ip] || []) {
        parts.push(`        <h3>nmap · ${esc(scan.tool)} · ${esc(scan.created_at)}</h3>\n`)
        if (scan.screenshots.length) {
          parts.push('        <div class="shots">\n')
          for (const shot of scan.screenshots) {
            parts.push(`            <div>
                <p class="meta" style="margin-bottom:5px">${esc(shot.url)}</p>
`)
            parts.push(shot.data_uri
              ? `                <img src="${shot.data_uri}" alt="screenshot of ${esc(shot.url)}">\n`
              : '                <p class="empty">Screenshot file missing.</p>\n')
            parts.push('            </div>\n')
          }
          parts.push('        </div>\n')
        }
        parts.push(`        <details>
            <summary>Command &amp; full nmap output</summary>
            <pre>${esc(scan.command)}

${esc(scan.raw_output)}</pre>
        </details>
`)
      }
      parts.push('    </div>\n')
    }
  }

  parts.push('</div>\n</body>\n</html>\n')
  return parts.join('')
}

// --- the two entry points -------------------------------------------------

function exportTask (taskId, format, version) {
  const task = store.getTask(taskId)
  if (!task) throw new Error('Task not found.')
  return render(format, {
    title: `${task.project_name} · ${task.name}`,
    slug: `task-${taskId}`,
    sections: [taskBundle(task)],
    version,
  })
}

function exportProject (projectId, format, version) {
  const project = store.getProject(projectId)
  if (!project) throw new Error('Project not found.')
  return render(format, {
    title: project.name,
    slug: `project-${projectId}`,
    sections: store.listTasks(projectId).map(taskBundle),
    project,
    version,
  })
}

module.exports = { taskBundle, render, exportTask, exportProject, renderText, renderHtml }
